be/src/exprs/map_functions.cpp

```diff
@@ -16,10 +16,111 @@
 
 #include "column/array_column.h"
 #include "column/map_column.h"
-#include "column/type_traits.h"
+#include "common/logging.h"
 
 namespace starrocks {
 
+// Used to construct a Map type value from keys and values. Input keys and values are in the
+// format of Array type.
+// For example,
+//  map([1, 2], [3, 4]) will generate a map which value is {1=3,2=4}
+StatusOr<ColumnPtr> MapFunctions::map(FunctionContext* context, const Columns& columns) {
+    DCHECK_EQ(2, columns.size());
+    RETURN_IF_COLUMNS_ONLY_NULL(columns);
+
+    auto& keys_column = columns[0];
+    NullColumn* keys_null = nullptr;
+    ArrayColumn* keys_data = nullptr;
+    if (keys_column->is_nullable()) {
+        auto keys = down_cast<NullableColumn*>(keys_column.get());
+        keys_null = keys->null_column().get();
+        keys_data = down_cast<ArrayColumn*>(ColumnHelper::get_data_column(keys));
+    } else {
+        keys_data = down_cast<ArrayColumn*>(keys_column.get());
+    }
+
+    auto& values_column = columns[1];
+    NullColumn* values_null = nullptr;
+    ArrayColumn* values_data = nullptr;
+    if (values_column->is_nullable()) {
+        auto values = down_cast<NullableColumn*>(values_column.get());
+        values_null = values->null_column().get();
+        values_data = down_cast<ArrayColumn*>(ColumnHelper::get_data_column(values));
+    } else {
+        values_data = down_cast<ArrayColumn*>(values_column.get());
+    }
+
+    auto& keys_offsets = keys_data->offsets().get_data();
+    auto& values_offsets = values_data->offsets().get_data();
+    auto num_rows = keys_data->size();
+    if (!keys_column->has_null() && !values_column->has_null()) {
+        size_t num_equals = 0;
+        // NOTE: only need to check the accumulated offset of each element
+        for (int i = 1; i <= num_rows; ++i) {
+            num_equals += (keys_offsets[i] == values_offsets[i]);
+        }
+        if (num_equals != num_rows) {
+            return Status::InvalidArgument("Key and value arrays must be the same length");
+        }
+        auto copied_key_elements = keys_data->elements().clone_shared();
+        auto copied_value_elements = values_data->elements().clone_shared();
+        auto copied_offsets = keys_data->offsets().clone_shared();
+        return MapColumn::create(std::move(copied_key_elements), std::move(copied_value_elements),
+                                 std::static_pointer_cast<UInt32Column>(copied_offsets));
+    } else {
+        // build the null column
+        NullColumnPtr null_column;
+        if (keys_null != nullptr) {
+            if (values_null != nullptr) {
+                null_column = std::static_pointer_cast<NullColumn>(keys_null->clone_shared());
+                ColumnHelper::or_two_filters(num_rows, null_column->get_data().data(), values_null->get_data().data());
```

Comment:
why `or_two_filters` for key column and value column?

Reply:
When either key or value is null, the result will be null.
For example:
[1, 2] and null will produce NULL map
NULL and [1] will product NULL map
So I compute an 'or' operation here to get the null-bits for the resulting map.